feat(lep): multiple backup stores support (#6630)
longhorn/longhorn: enhancements/20240926-multiple-backup-targets-support.md

https://github.com/longhorn/longhorn/issues/5411

## Motivation

Users can choose backup targets to that they want to back up volumes by setting the backup target name of the volumes.

### Goals

- There are different backup targets at the same time on Longhorn.
- Backups can be synchronized from different remote backup targets.
- Snapshots of a volume can be backed up to a specific remote backup target according to the backup target name of the volume.
- A backup can be restored from its remote backup targets to be a new volume.

### Non-goals [optional]

- Snapshots of a volume can not be backed up to different remote backup targets concurrently.

## Proposal

1. Allow users to create, delete or modify backup targets.
    - The backup target controller needs to synchronize the backups' information from the remote backup target and create `Backup` CR objects and `BackupVolume` CR objects in the cluster when the backup target is created.
    - When the backup target is deleted, backup target controller needs to delete the `Backup` CR and `BackupVolume` CR objects in the cluster related to the backup target, and backup data will be kept on the remote backup target.

> 2. Introduce new fields `ReadOnly` in `Spec` of `BackupTarget` CRD.

Reviewer: Why do we need `readonly`? is it part of the multiple backup target story?

Author: Added the user story. @james-munson recommended it in https://github.com/longhorn/longhorn/pull/6630#discussion_r1316370624.
Users can only read data from the backup target and can not store the data on the backup target for certain reasons (such as the remote backup target server is running out of storage).

Reviewer: This might be over-consideration.  We can add this to No-Goals first.

5. Introduce new fields `BackupTargetName` and `BackingImage` in `Spec` of `BackupBackingImage` CRD to keep information of the remote backup target and backing image name.
    - Different backup targets might have backing images with the same names.
6. Introduce a new field `BackupTargetName` in `Spec` of the `RecurringJob` CRD to keep backup target information.
7. Introduce a new field `BackupTargetName` in `Spec` of the `Volume` CRD to mark the default backup target for storing the volume backups.
8. The system backup will be handled by the default backup target.

### User Stories

User need to create/modify backup targets on the page `Setting/Backup Target` as `Setting/Backing Image` via the Longhorn UI or by the manifest. After backup targets are created and synchronized, users can start to create a backup on these remote backup targets.

### User Experience In Detail

#### Create Or Modify Backup Targets By UI

1. User can find the link `Backup Target` in drop down menu of `Setting`.
2. The page `Backup Target` would show backup target name, backup target URL, secret name (credential) for this backup target, poll interval, and availability status.

> 3. There will be an empty default backup target created by Longhorn.

Reviewer: There seems no user story about setting the default backup target mentioned here. Users should be able to set the default backup target from the available backup targets.

Author: Added the user story for the default backup target.
The LEP does not allow users to set the default backup target, and the default backup target will always be the backup target `default` that Longhorn will create.

5. Users can modify a backup target information by clicking the button `Edit`.
6. Users can set the backup target `ReadOnly` true and then users can not create any backups on it.
7. Users can delete a backup target by clicking the button `Delete`.
8. Users can not create or modify a backup target with the `BackupTargetURL` that is the same to an existing backup target.
9. Users can not delete the default backup target directly.
10. Users can not modify the backup target name.
11. Users can not modify the `BackupTargetURL` of a backup target that is the same to another one in the cluster.

#### Create Or Modify Backup Targets By CLI

User can the bash command to create or modify a backup target such as below:

```bash
cat <<EOF >>new-backup-target.yaml
apiVersion: longhorn.io/v1beta2
kind: BackupTarget
metadata:
  name: azure-blob-server-001
  namespace: longhorn-system
spec:
  backupTargetURL: azblob://demo@core.windows.net/
  credentialSecret: azblob-secret
  pollInterval: 4m30s
status:

>   readOnly: false

Reviewer: What does `readOnly` in status mean? I thought it's in the spec if we want to have it.

Author: Yes, it is sufficient in the `spec`, and removed it from the `status`. Thanks.

EOF

kubectl apply -f new-backup-target.yaml
```

#### Create A Backup To Remote Backup Targets By UI

After users click the button `Create Backup` on the `Volume Details` page, the backup will be stored on the remote backup target set in the `BackupTargetName` filed of the volume.

#### Create A Backup To Remote Backup Targets By CLI

1. Create a snapshot of the volume first.

> 2. Create backups to store the snapshot by the manifest as below:

Reviewer: ```suggestion
2. Create a backup referring to the snapshot by the manifest as below:
```

Author: Modified. Thanks.

```yaml
apiVersion: longhorn.io/v1beta2
kind: Backup
metadata:
  labels:
    backup-volume: pvc-xxxxxxxx-9ab9-4055-9eb4-558999b09a11
  name: backup-test-001
  namespace: longhorn-system
spec:
  backupTargetName: azure-blob-server-001 # If this field is assigned, it should be the same to the `spec.backupTargetName` field in the volume of the snapshot.
  labels:
    longhorn.io/volume-access-mode: rwo
  snapshotName: xxxxxxxx-9a69-4ede-ab88-c9853459462c
```

> 3. If the field `spec.backupTargetName` is empty, it will be replaced by the field `spec.backupTargetName` in the volume of the snapshot.

Reviewer: It may be helpful to clarify how `backupTargetname` is prioritized. For instance, a backup takes precedence over a volume, while the lowest priority is assigned to the backup target set globally.

Author: The field. `Spec.BackupTargetName` of the `Backup` is removed.
Added the description for the priority in the section `Default Backup Target` of the user stories.

> 4. If the `spec.backupTargetName` field will be assigned, it should be the same to the `spec.backupTargetName` field in the volume of the snapshot.

Reviewer: If a snapshot’s `backupTargetName` should always match its volume, why do we need this specific field in Backup CRD instead of using the value directly from the volume?

Author: Yes, you are right. I removed the filed `Spec.BackupTargetName` of the `Backup` and moved it to the `Status` to specify the backup target of the `Backup`. Thanks

#### Create A Backup To Remote Backup Targets By Recurring Jobs

It will store snapshots to the appointed backup target set in the `spec.backupTargetName` of the volume when the recurring job is running.

#### Restore From A Backup From A Remote Backup Target

Restoring from a backup will behave as before. Choose a backup and do the `Restore` operation in drop down menu on UI.

> #### Create A Disaster Recovery Volume From A Remote Backup Target By UI

Reviewer: Is DR available for multiple backup targets? or just for the default global backup target?

Author: It should be available for multiple backup targets, do you think we should limit it to the default global backup target?

Reviewer: No. If it's feasible, it's good to have. We don't need to do any differentiation here.

   - Remove a backup volume or backup backing image CR object if the corresponding backup target is deleting or the URL of the corresponding backup target is empty.
2. Modify the backup volume controller to tell the backup volume belongs to which backup target and synchronize the backups of the backup volume from remote backup target.
   - Create pulling backups from the backup target with the field `Spec.backupTargetName` of the backup target.
   - Clean up backup CR objects in the cluster by the volume name and backup target name when deleting a backup volume or the backup target URL is empty.
3. Modify backup backing image controller
   - Start backing up a backing image with the given backup target name.

### Validating And Mutating Webhook

- Fill in the `Spec.BackupTargetName` field of the `Backup` CR object by the `Spec.BackupTargetName` filed in the `Volume` CR object if empty.
- Validate the `Spec.BackupTargetName` field of the `Backup` CR object is the same to the `Spec.BackupTargetName` field in the `Volume` CR object
- Validate the backup target is read-only or not when creating a backup.

> ### Test plan

Reviewer: Suggest adding test coverage for DR and updating the global backup target.

Author: Added test cases for updating the default global backup tarter and creating a DR volume. Thanks.

    - When the backup target is deleted, backup target controller needs to delete the `Backup` CR and `BackupVolume` CR objects in the cluster related to the backup target, and backup data will be kept on the remote backup target.
2. Introduce a new field `BackupTarget` in `Status` of the `Backup` CRD to keep backup target information.
    - The field `BackupTarget:` is used to specify the backup target of the backup for the backup information from the `kubectl get` command and Longhorn RESTful APIs.
3. Introduce new fields `BackupTargetName` and `VolumeName` in `Spec` of `BackupVolume` CRD to keep information of the remote backup target and volume name.
4. Introduce new fields `BackupTargetName` and `BackingImage` in `Spec` of `BackupBackingImage` CRD to keep information of the remote backup target and backing image name.
5. Introduce a new field `BackupTargetName` in `Spec` of the `Volume` CRD to mark the default backup target for storing the volume backups.
6. The system backup will be handled by the default backup target.

### User Stories

> #### Default Backup Target

Reviewer 2: I believe we are removing the backup target-related settings on the Setting page, right? Could you include this info in the LEP?

Author: No, we are not removing the backup target-related setting.
The default backup target will be able to be modified by the backup target-related settings.

> The `default` backup target will be created by Longhorn after installing Longhorn, and it will be the default backup target when creating a volume and a system backup. Users need to set up the `default` backup target before creating a new backup target. The settings `backup-target`, `backup-target-credential-secret`, and `backupstore-poll-interval` are still used to set up the `default` backup target.

Reviewer 2: Consider clarify on what happens if a user attempts to delete the `default` backup target.

Author: Added a description prohibiting the deletion of the `default` backup target.

> When creating a volume with an empty `BackupTagetName` filed, this filed will be populated with the default backup target name `default`.

Reviewer 2: ```suggestion
When creating a volume with an empty `BackupTargetName` field, this field will be populated with the default backup target name `default`.
```

Author: Modified. Thanks

8. Users can not delete the default backup target directly.
9. Users can not modify the backup target name.
10. Users can not modify the `BackupTargetURL` of a backup target that is the same to another one in the cluster.

#### Create Or Modify Backup Targets By CLI

User can the bash command to create or modify a backup target such as below:

```bash
cat <<EOF >>new-backup-target.yaml
apiVersion: longhorn.io/v1beta2
kind: BackupTarget
metadata:
  name: azure-blob-server-001
  namespace: longhorn-system
spec:
  backupTargetURL: azblob://demo@core.windows.net/
  credentialSecret: azblob-secret
  pollInterval: 4m30s
EOF

kubectl apply -f new-backup-target.yaml
```

#### Backup Target Conditions

There is only one backup target condition and one reason now.

```golang
const (
	BackupTargetConditionTypeUnavailable = "Unavailable"
	BackupTargetConditionReasonUnavailable = "Unavailable"
)
```

The condition `BackupTargetConditionTypeUnavailable` will be set to true if the backup target URL is empty or if retrieving backup information from the remote backup target fails due to issues such as a timeout or an unreachable backup target.

> #### Create A Backup To Remote Backup Targets By UI

Reviewer 2: How will the Backup page display backups from different targets?

Author: Added a description for displaying the different backup targets on the `Backup` page.

```txt
On the `Backup` page, the backup targets are displayed first if one or more backup volumes exist. Clicking a backup target will expand or collapse the list to show the associated backup volumes.
```

    }
  ```

  ```yaml
  metadata:
    labels:
      longhorn.io/backup-target: the backup target name (string)
      longhorn.io/backup-volume: the volume name (string)
    name: the backup volume name. (string)
  spec:
    backupTargetName: the backup target Name. (string)
    volumeName: the volume name (string)
    ...
  ```

3. Modify the Backup Backing Image CRD `backupbackingimages.longhorn.io` to add new fields `BackingImage` and `BackupTargetName` in `Spec`, and add labels `LonghornLabelBackupTarget = "backup-target"` and `LonghornLabelBackingImage = "backing-image"`.

  ```golang
  type BackupBackingImageSpec struct {
    ...
    // The backing image name.
    BackingImage string `json:"backingImage"`
    // The backup target name.
    BackupTargetName string `json:"backupTargetName"`
  }
  ```

  ```yaml
  metadata:
    labels:
      longhorn.io/backing-image: the backing image name (string)
      longhorn.io/backup-target: the backup target name (string)
    name: the backup backing image name. (string)
  spec:
    backingImage: the backing image name (string)
    backupTargetName: the backup target Name. (string)
    ...
  ```

> 4. Modify the Volume CRD `volumes.longhorn.io` to add a new field `BackupTargetName` in `Spec` for a default backup target.

Reviewer 2: Can the volume's `BackupTargetName` be updated? If so, how will Longhorn handle it?

Author: Yes, it could be updated. If it is updated by another backup target name, the next backups to be created will be stored on the new backup target.

   - Create pulling backup backing images with a random backup volume name, for example, `bbi-xxxxxx-a12345-b12345-xxxxxx` from the backup target with filling in the fields `Spec.BackupTargetName` and `Spec.BackingImage`, and corresponding labels if not existing.
   - Remove a backup volume or backup backing image CR object if the corresponding backup target is deleting or the URL of the corresponding backup target is empty.
2. Modify the backup volume controller to tell the backup volume belongs to which backup target and synchronize the backups of the backup volume from remote backup target.
   - Create pulling backups from the backup target with the field `Status.BackupTarget` of the backup target.
   - Clean up backup CR objects in the cluster by the volume name and backup target name when deleting a backup volume or the backup target URL is empty.
3. Modify backup backing image controller
   - Start backing up a backing image with the given backup target name.

### Validating And Mutating Webhook

- If the `Spec.BackupTargetName` field of the Volume CR object is empty, populate it with the default backup target name.

> ### Test plan

Reviewer 2: We should add test coverage for the backup and restore of volumes in the system backup and restore process.

Author: Added. Thanks.

# Multiple Backup Targets Support

## Summary

The current Longhorn only support a single backup target for all volumes. The feature aims to provide multiple backup targets.

### Related Issues

https://github.com/longhorn/longhorn/issues/2317
https://github.com/longhorn/longhorn/issues/5411

## Motivation

Users can choose backup targets to that they want to back up volumes by setting the backup target name of the volumes.

### Goals

- There are different backup targets at the same time on Longhorn.
- Backups can be synchronized from different remote backup targets.
- Snapshots of a volume can be backed up to a specific remote backup target according to the backup target name of the volume.

> - A backup can be restored from its remote backup targets to be a new volume.

Reviewer 3: Single backup target for each volume

```suggestion
- A backup can be restored from its remote backup target to be a new volume.
```

Author: Modified. Thanks.

### Non-goals [optional]

- Snapshots of a volume can not be backed up to different remote backup targets concurrently.

## Proposal

1. Allow users to create, delete or modify backup targets.
    - The backup target controller needs to synchronize the backups' information from the remote backup target and create `Backup` CR objects and `BackupVolume` CR objects in the cluster when the backup target is created.
    - When the backup target is deleted, backup target controller needs to delete the `Backup` CR and `BackupVolume` CR objects in the cluster related to the backup target, and backup data will be kept on the remote backup target.

> 2. Introduce new fields `ReadOnly` in `Spec` of `BackupTarget` CRD.

Reviewer 3: Let's remove `ReadOnly` which is out of scope of the feature.
```suggestion
2. Introduce new fields `ReadOnly` in `Spec` of `BackupTarget` CRD.
```

Author: Removed. Thanks

    - `ReadOnly` marks that the remote backup target is read-only and users can not create any backups on it.
3. Introduce a new field `BackupTarget` in `Status` of the `Backup` CRD to keep backup target information.
    - The field `BackupTarget:` is used to specify the backup target of the backup.
4. Introduce new fields `BackupTargetName` and `VolumeName` in `Spec` of `BackupVolume` CRD to keep information of the remote backup target and volume name.

>     - Different backup targets might have volumes with the same names.

Reviewer 3: Only show one backupvolume?

Author: Removed the confusing statement.

>Only show one backupvolume?

No. It will be one-to-one

- volume A on backup target A
```yaml
 metadata:
   name: the backup volume name. (random string)
 spec:
   backupTargetName: backupTargetA
   volumeName: volumeA
   ...
```
- volume A on backup target B
```yaml
 metadata:
   name: the backup volume name. (random string)
 spec:
   backupTargetName: backupTargetB
   volumeName: volumeA
   ...
```

Reviewer 3: This is important. Please add the information to the LEP. Thank you.

Author: Added it back. Thanks

6. Introduce a new field `BackupTargetName` in `Spec` of the `Volume` CRD to mark the default backup target for storing the volume backups.
7. The system backup will be handled by the default backup target.

### User Stories

#### Default Backup Target

The `default` backup target will be created by Longhorn after installing Longhorn, and it will be the default backup target when creating a volume and a system backup. Users need to activate the `default` backup target before creating a new backup target. The settings `backup-target`, `backup-target-credential-secret`, and `backupstore-poll-interval` are still used to set up the `default` backup target.

When creating a volume with an empty `BackupTagetName` filed, this filed will be populated with the default backup target name `default`.

#### Create a new backup target

Users need to create/modify backup targets on the page `Setting/Backup Target` as `Setting/Backing Image` via the Longhorn UI or by the manifest. After backup targets are created and synchronized, users can start to create a backup on these remote backup targets.

#### A Read-Only backup target

> When users attempt to restore existing backups from a remote backup target that is read-only or cannot create a new backup for certain reasons (such as insufficient storage), they can set the `Spec.ReadOnly` field to true for this backup target.

Reviewer 3: Suggest removing the read-only feature. It is out of scope.

Author: Removed. Thanks

### User Experience In Detail

#### Create Or Modify Backup Targets By UI

> 1. There will be an empty default backup target created by Longhorn and users should activate the default backup target before creating a new backup target.

Reviewer 3: What does `activate` mean?

Author: It means `set up` and the backup target is available.
Modified it to be `set up`. Thanks.

2. User can find the link `Backup Target` in drop down menu of `Setting`.
3. The page `Backup Target` would show backup target name, backup target URL, secret name (credential) for this backup target, poll interval, and availability status.
4. User can create a new backup target on the page `Backup Target` by clicking the button `Create`. User have to fill out the necessary item for the backup target name, and it can be optional for the backup target URL, secret name, poll interval, and read-only fields.
5. Users can modify a backup target information by clicking the button `Edit`.

> 6. Users can set the backup target `Spec.ReadOnly` true and then users can not create any backups on it.

Reviewer 3: Suggest removing it.

Author: Removed. Thanks

Author: Removed. Thanks

7. Users can delete a backup target by clicking the button `Delete`.
8. Users can not create or modify a backup target with the `BackupTargetURL` that is the same to an existing backup target.
9. Users can not delete the default backup target directly.
10. Users can not modify the backup target name.
11. Users can not modify the `BackupTargetURL` of a backup target that is the same to another one in the cluster.

> 

Reviewer 3: Will you introduce `backupTarget.status.conditions` to help understand the status and troubleshoot?

Reviewer 3: Another question: 
If volume `abc`''s backups are both in `default`, `backupTarget A` and `backupTarget B`, what will Longhorn show in the two cases

- `volume.spec.backupTargetName` is `""`
- `volume.spec.backupTargetName` is `"default"`
- `volume.spec.backupTargetName` is `"backupTarget A"`

Author: >Will you introduce backupTarget.status.conditions to help understand the status and troubleshoot?

Added a section in user stories for backup target conditions.

>Another question:
If volume abc''s backups are both in default, backupTarget A and backupTarget B, what will Longhorn show in the two cases

Three backup CR objects will be created corresponding to the backup targets

- backup target "default"
```yaml
metadata:
  labels:
    longhorn.io/backup-target: defaultBackupTagetName
  name: the backup name. (random string)
status:
  backupTarget: defaultBackupTagetName
  volumeName: abc
  ...
```
- backup target A
```yaml
metadata:
  labels:
    longhorn.io/backup-target: backupTagetAName
  name: the backup name. (random string)
status:
  backupTarget: backupTagetAName
  volumeName: abc
  ...
```
- backup target B
```yaml
metadata:
  labels:
    longhorn.io/backup-target: backupTagetBName
  name: the backup name. (random string)
status:
  backupTarget: backupTagetBName
  volumeName: abc
  ...
```

The differences are the `Status.BackupTarget` field and `longhorn.io/backup-target` label.

>- volume.spec.backupTargetName is ""

It will be not be empty, and it would be populated with the default backup target name by the volume mutator when creating.

>- volume.spec.backupTargetName is "default"
>- volume.spec.backupTargetName is "backupTarget A"

The `volume.spec.backupTargetName` is irrelated to the backup after the backup is created.

Reviewer 3: `defaultBackupTagetName` => what's this? `default`?

> It will be not be empty, and it would be populated with the default backup target name by the volume mutator when creating.

How to handle the existing backups?

Author: >defaultBackupTagetName => what's this? `default`?

Yes, the default backup target name is `default`.

>How to handle the existing backups?

In the section `Upgrade strategy`, it will fill in the labels `longhorn.io/backup-target` and field `BackupTarget` in `Status` for existing Backup CR objects with  the `default` backup target name.

Reviewer 3: One question: why do we need `status.BackupTarget`?

Author: The field `status.BackupTarget` is used to point out this backup is stored on (synced from) this backup target.

Reviewer 3: I see. Why not replace ` longhorn.io/backup-target` with `spec.BackupTarget`

Author: >I see. Why not replace  longhorn.io/backup-target with spec.BackupTarget

Yes, we could use the label `longhorn.io/backup-target` instead of the field `Status.BackupTarget`, but users can easily modify the label `longhorn.io/backup-target` and it will be complex when we get the backup target name for the information shown on CLI columns or returned by RESTFul APIs.

Reviewer 3: No, I mean `Spec.BackupTarget`. What's the purpose of the label? Why not use spec field?

Author: >No, I mean `Spec.BackupTarget`.

Not sure which filed you mentioned, do you mean the field `Volume.Spec.BackupTargetName` or `Backup.Spec.BackupTarget`?

If you meant the field `backup.Spec.BackupTarget`, I removed it from the `Spec` in the LEP and moved it to the field `Backup.Status.BackupTarget`.

>What's the purpose of the label? Why not use spec field?

The label is used to get the backups of the backup target and volume quickly.

Reviewer 3: > Not sure which filed you mentioned, do you mean the field Volume.Spec.BackupTargetName or Backup.Spec.BackupTarget?

I mean why not remove the label and use Spec.BackupTarget instead.

> The label is used to get the backups of the backup target and volume quickly.

OK. Got it. 

Do we really need `Status.BackupTargetName`? We already have the label. Isn't it?

Author: Added some descriptions for the field `Status.BackupTarget`. Thanks.

# Multiple Backup Targets Support

## Summary

> The current Longhorn only support a single backup target for all volumes. The feature aims to provide multiple backup targets.

Reviewer 3: ```suggestion
The current Longhorn only supports a single backup target for all volumes. The feature aims to provide multiple backup targets.
```

Author: Modified. Thanks

### Related Issues

https://github.com/longhorn/longhorn/issues/2317
https://github.com/longhorn/longhorn/issues/5411

## Motivation

Users can choose backup targets to that they want to back up volumes by setting the backup target name of the volumes.

### Goals

- There are different backup targets at the same time on Longhorn.

> - Backups can be synchronized from different remote backup targets.

Reviewer 4: Are we addressing this goal in this LEP?
I don't see the implementation design for this one in the following docs

Author: When a new backup target is created and available, the backups on this remote backup target will be synchronized immediately by the backup target, backup volume, and backup controllers.

Reviewer 4: Got it.

- Modify the API `BackupBackingImageCreate`, `BackupBackingImageGet` and `BackupBackingImageList`:
  - Add new fields `BackingImage` and `BackupTargetName` in the input `BackupBackingImage`.

  ```golang
    type BackupBackingImage struct {
      client.Resource

      Name             string `json:"name"`
      BackingImage     string `json:"backingImage"`
      BackupTargetName string `json:"backupTargetName"`
      ...
    }
  ```

## Design

### Implementation Overview

#### Custom Resource Definitions

1. Modify the Backup CRD `backups.longhorn.io` to add a new field `BackupTarget` in `Status` for creating a backup. And add a label `LonghornLabelBackupTarget = "backup-target"` for resource listing.

  ```golang
  type BackupStatus struct {
    ...
    // The backup target name.
    BackupTarget string `json:"backupTarget"`
  }
  ```

  ```yaml
  metadata:
    labels:
      longhorn.io/backup-target: the backup target name (string)
      longhorn.io/backup-volume: the volume name (string)
    name: the backup name. (string)
  status:
    backupTarget: the backup target Name. (string)
    ...
  ```

> 2. Modify the Backup Volume CRD `backupvolumes.longhorn.io` to add new fields `BackupTargetName` and `VolumeName` in `Spec`, and add labels `LonghornLabelBackupTarget` and `LonghornLabelBackupVolume` for resource listing.

Reviewer 4: Does that means we will have >= 1 BackupVolumes for 1 Volume ?
But I thought the BackupVolume's name is the same as the Volume without any random string or UUID
So how do we create multiple BackupVolume CR with the same name in the cluster?

Author: >Does that means we will have >= 1 BackupVolumes for 1 Volume ?

Yes, identifying the backup volume depends on the combination of the backup target name and volume name.

The backup volume pulled from the remote backup target will have a [random backup volume name](https://github.com/longhorn/longhorn/pull/6630/files#diff-f66bf27d07bfda116e6936671071bf1ef0ce469653a32bdb0c1c3571f8a6c1b5R273).

Reviewer 4: Oh I see the implementation in the code.
Got it, thanks

  ```golang
    type BackupVolumeSpec struct {
      ...
      SyncRequestedAt metav1.Time `json:"syncRequestedAt"`
      // The backup target name that the backup volume was synced.
      BackupTargetName string `json:"backupTargetName"`
      // The volume name that the backup volume was used to backup.
      VolumeName string `json:"volumeName"`
    }
  ```

  ```yaml
  metadata:
    labels:
      longhorn.io/backup-target: the backup target name (string)
      longhorn.io/backup-volume: the volume name (string)
    name: the backup volume name. (string)
  spec:
    backupTargetName: the backup target Name. (string)
    volumeName: the volume name (string)
    ...
  ```

> 3. Modify the Backup Backing Image CRD `backupbackingimages.longhorn.io` to add new fields `BackingImage` and `BackupTargetName` in `Spec`, and add labels `LonghornLabelBackupTarget = "backup-target"` and `LonghornLabelBackingImage = "backing-image"`.

Reviewer 4: Same as the Volume, since the BackupBackingImage is using the same name as the BackingImage
If users creates two BackupBackingImage from the same BackingImage with different backup targets
How to avoid the name conflict?

# Multiple Backup Targets Support

## Summary

The current Longhorn only supports a single backup target for all volumes. The feature aims to provide multiple backup targets.

### Related Issues

https://github.com/longhorn/longhorn/issues/2317
https://github.com/longhorn/longhorn/issues/5411

## Motivation

Users can choose backup targets to that they want to back up volumes by setting the backup target name of the volumes.

### Goals

- There are different backup targets at the same time on Longhorn.
- Backups can be synchronized from different remote backup targets.
- Snapshots of a volume can be backed up to a specific remote backup target according to the backup target name of the volume.
- A backup can be restored from its remote backup target to be a new volume.

### Non-goals [optional]

> - Snapshots of a volume can not be backed up to different remote backup targets concurrently.

Author: Yes, as you mentioned. Modified. Thanks.

## Proposal

1. Allow users to create, delete or modify backup targets.
    - The backup target controller needs to synchronize the backups' information from the remote backup target and create `Backup` CR objects and `BackupVolume` CR objects in the cluster when the backup target is created.
    - When the backup target is deleted, backup target controller needs to delete the `Backup` CR and `BackupVolume` CR objects in the cluster related to the backup target, and backup data will be kept on the remote backup target.

> 2. Introduce a new field `BackupTarget` in `Status` of the `Backup` CRD to keep backup target information.

Author: Modified. Thanks and I will remember next time.

    - The field `BackupTarget:` is used to specify the backup target of the backup for the backup information from the `kubectl get` command and Longhorn RESTful APIs.
3. Introduce new fields `BackupTargetName` and `VolumeName` in `Spec` of `BackupVolume` CRD to keep information of the remote backup target and volume name.
    - Different backup targets might have volumes with the same names. The `BackupVolume` object name will be a randomly generated UUID with a prefix. It uses the `BackupTargetName` and `VolumeName` fields to specify the volume on the remote backup target.

> 4. Introduce new fields `BackupTargetName` and `BackingImage` in `Spec` of `BackupBackingImage` CRD to keep information of the remote backup target and backing image name.

Author: >If the implementation is straightforward, feel free to proceed, but it's nice to have.

Yes, it is implemented.

    - Different backup targets might have backing images with the same names. The `BackupBackingImage` object name will be a randomly generated UUID with a prefix. It uses the `BackupTargetName` and `BackingImage` fields to specify the backing image on the remote backup target.
5. Introduce a new field `BackupTargetName` in `Spec` of the `Volume` CRD to mark the default backup target for storing the volume backups.

> 6. The system backup will be handled by the default backup target.

Author: Modified. Thanks

7. Support the parameter `backupTargetName` in the `StorageClass`.

> ### User Stories

Reviewer 3: Add some cases
- Case1
	- Create a storageClass without backupTargetName
	- Create a volume that volume.spec.backupTargetName will be assigned
- Case2
	- Create a storageClass with backupTargetName
	- Create a volume that volume.spec.backupTargetName will be assigned
- Case3
	- Delete backupTarget when DR volumes are existing

Reviewer: >Delete backupTarget when DR volumes are existing

Is this a valid case? I thought DR volumes only from the default backup target which should not be deleted.

Author: >Add some cases
>- Case1
>-- Create a storageClass without backupTargetName
>-- Create a volume that volume.spec.backupTargetName will be assigned
>- Case2
>-- Create a storageClass with backupTargetName
>-- Create a volume that volume.spec.backupTargetName will be assigned
>- Case3
>-- Delete backupTarget when DR volumes are existing

@derekbit Added. Thanks.

>Is this a valid case? I thought DR volumes only from the default backup target which should not be deleted.

Yes, we can create DR volumes from all backup targets.

> #### Default Backup Target

Author: Modified. Thanks.

>Additionally, DR volumes will only function as expected when using the default backup target.

We can use all backup targets created in the cluster.

The `default` backup target will be created by Longhorn after installing Longhorn, and it will be the default backup target when creating a volume and a system backup. Users need to set up the `default` backup target before creating a new backup target. The settings `backup-target`, `backup-target-credential-secret`, and `backupstore-poll-interval` are still used to set up the `default` backup target, and the setting will be reviewed and updated if the `default` backup target CR is modified.

The `default` backup target can not be deleted by users.

When creating a volume with an empty `BackupTargetName` filed, this filed will be populated with the default backup target name `default`.

#### Create a new backup target

> Users need to create/modify backup targets on the page `Setting/Backup Target` as `Setting/Backing Image` via the Longhorn UI or by the manifest. After backup targets are created and synchronized, users can start to create a backup on these remote backup targets.

Reviewer: When deleting a backup target which is not default, suggest considering the following checks. 
- For existing volumes (runtime), if it's referenced by volumes, they can't be deleted until the volumes don't use it.
- If it has been used by existing storage classes, a warning notification (log, event) should be created. Also, users should be warned when deleting from UI. 
- When creating new volumes (creation time), as the backup target is deleted, so the volumes should not be created and the corresponding error education should be clear for users.

Author: >For existing volumes (runtime), if it's referenced by volumes, they can't be deleted until the volumes don't use it.

We left the update operation (the filed `Volume.Spec.BackupTargetName`) to users if they try to delete the backup target referenced by volumes.

>If it has been used by existing storage classes, a warning notification (log, event) should be created. Also, users should be warned when deleting from UI.

Yes, a warning/event will be helpful. Added in the `Create/Delete Backup Target` section.

>When creating new volumes (creation time), as the backup target is deleted, so the volumes should not be created and the corresponding error education should be clear for users.

Now the implementation only check if the backup target exists, and yes, we should check if the backup target is deleting.

### User Experience In Detail

#### Create Or Modify Backup Targets By UI

1. There will be an empty default backup target created by Longhorn and users should set up the default backup target before creating a new backup target.
2. User can find the link `Backup Target` in drop down menu of `Setting`.
3. The page `Backup Target` would show backup target name, backup target URL, secret name (credential) for this backup target, poll interval, and availability status.
4. User can create a new backup target on the page `Backup Target` by clicking the button `Create`. User have to fill out the necessary item for the backup target name, and it can be optional for the backup target URL, secret name, and poll interval fields.
5. Users can modify a backup target information by clicking the button `Edit`.

> 6. Users can delete a backup target by clicking the button `Delete`.

Reviewer: A webhook validation is needed to check the conditions that prevent deletion. This is also applied to creation/modification cases.

Author: It only prevents deleting the backup target when DR volumes exist now

7. Users can not create or modify a backup target with the `BackupTargetURL` that is the same to an existing backup target.
8. Users can not delete the default backup target directly.
9. Users can not modify the backup target name.
10. Users can not modify the `BackupTargetURL` of a backup target that is the same to another one in the cluster.

#### Create Or Modify Backup Targets By CLI

User can the bash command to create or modify a backup target such as below:

```bash
cat <<EOF >>new-backup-target.yaml
apiVersion: longhorn.io/v1beta2
kind: BackupTarget
metadata:
  name: azure-blob-server-001
  namespace: longhorn-system
spec:
  backupTargetURL: azblob://demo@core.windows.net/
  credentialSecret: azblob-secret
  pollInterval: 4m30s
EOF

kubectl apply -f new-backup-target.yaml
```

#### Backup Target Conditions

There is only one backup target condition and one reason now.

```golang
const (
	BackupTargetConditionTypeUnavailable = "Unavailable"
	BackupTargetConditionReasonUnavailable = "Unavailable"
)
```

> The condition `BackupTargetConditionTypeUnavailable` will be set to true if the backup target URL is empty or if retrieving backup information from the remote backup target fails due to issues such as a timeout or an unreachable backup target.

Reviewer: Suggest adding more reasoning to the condition message, so users can better understand the cause of the unavailability.

Author: Added some error reasons we encountered.

2. Create a backup referring to store the snapshot by the manifest as below:

```yaml
apiVersion: longhorn.io/v1beta2
kind: Backup
metadata:
  labels:
    backup-volume: pvc-xxxxxxxx-9ab9-4055-9eb4-558999b09a11
  name: backup-test-001
  namespace: longhorn-system
spec:
  labels:
    longhorn.io/volume-access-mode: rwo
  snapshotName: xxxxxxxx-9a69-4ede-ab88-c9853459462c
```

3. The `status.backupTarget` field will be assigned to match the `spec.backupTargetName` field in the volume of the snapshot after the backup is completed.  
  The field `status.backupTarget` displays the backup target name in the column information of the backup when users run the `kubectl get` command, and populates the backup target name in the backup information returned by the Longhorn RESTful APIs.

#### Create A Backup To Remote Backup Targets By Recurring Jobs

It will store snapshots to the appointed backup target set in the `spec.backupTargetName` of the volume when the recurring job is running.

#### Restore From A Backup From A Remote Backup Target

Restoring from a backup will behave as before. Choose a backup and do the `Restore` operation in drop down menu on UI.

> #### Create A Disaster Recovery Volume From A Remote Backup Target By UI

Reviewer: If we create DR volumes for volumes that are using multiple backup targets, how does DR work in the remote cluster?

I assume the remote cluster uses only the default backup target for DR volumes. In this case, will the above DR volumes incrementally restore to the remote backup target? Or will users need to configure all backup targets in the remote cluster to ensure all DR volumes are incrementally restored?

Suggest explaining this clear.

Author: Added some explanations. Thanks.

    | Get    |  | backupTarget BackupTarget, err error | Get the backup targets information | **GET** `/v1/backuptargets/{backupTargetName}` |

- Modify the APIs `BackupVolumeGet`, `BackupVolumeList`:
  - Add a new field `BackupTargetName` and `VolumeName` in returning `BackupVolume` information for all backup targets that have this backup volume.
- Modify the APIs `BackupGet`:
  - Add a new field `BackupTargetName` in returning `Backup` information.
- Modify the API `SnapshotBackup`:
  - Add a new field `BackupTargetName` in the input `SnapshotInput` and returning `Backup` information.
- Modify the API `BackupBackingImageCreate`, `BackupBackingImageGet` and `BackupBackingImageList`:
  - Add new fields `BackingImage` and `BackupTargetName` in the input `BackupBackingImage`.

  ```golang
    type BackupBackingImage struct {
      client.Resource

      Name             string `json:"name"`
      BackingImage     string `json:"backingImage"`
      BackupTargetName string `json:"backupTargetName"`
      ...
    }
  ```

## Design

> ### Implementation Overview

Reviewer 3: Add
1. How to handle backupTarget CR deletion
  - backup CRs handling
  - backupVolume CRs handling
  - volume CRs handling
  - DR volumes handling

2. How to handle the the backupTarget CRs while uninstalling Longhorn system

Author: Added. Thanks.

> #### Custom Resource Definitions

Reviewer 3: Need to add the ownerReference Design

Author: Added Thanks.

   - Remove a backup volume or backup backing image CR object if the corresponding backup target is deleting or the URL of the corresponding backup target is empty.
2. Modify the backup volume controller to tell the backup volume belongs to which backup target and synchronize the backups of the backup volume from remote backup target.
   - Create pulling backups from the backup target with the field `Status.BackupTarget` of the backup target.
   - Clean up backup CR objects in the cluster by the volume name and backup target name when deleting a backup volume or the backup target URL is empty.
3. Modify backup backing image controller
   - Start backing up a backing image with the given backup target name.

### Validating And Mutating Webhook

- If the `Spec.BackupTargetName` field of the Volume CR object is empty, populate it with the default backup target name.
- The volume validator will check if the backup target exists after the volume is created or the field `Spec.BackupTargetName` is updated.

> ### Test plan

Reviewer: Miss backup image test cases if you want to add this functionality.

Author: Added. Thanks.

  spec:
    backupTargetName: the backup target Name. (string)
    volumeName: the volume name (string)
    ...
  ```

3. Modify the Backup Backing Image CRD `backupbackingimages.longhorn.io` to add new fields `BackingImage` and `BackupTargetName` in `Spec`, and add labels `LonghornLabelBackupTarget = "backup-target"` and `LonghornLabelBackingImage = "backing-image"`.

  ```golang
  type BackupBackingImageSpec struct {
    ...
    // The backing image name.
    BackingImage string `json:"backingImage"`
    // The backup target name.
    BackupTargetName string `json:"backupTargetName"`
  }
  ```

  ```yaml
  metadata:
    labels:
      longhorn.io/backing-image: the backing image name (string)
      longhorn.io/backup-target: the backup target name (string)
    name: the backup backing image name. (string)
  spec:
    backingImage: the backing image name (string)
    backupTargetName: the backup target Name. (string)
    ...
  ```

4. Modify the Volume CRD `volumes.longhorn.io` to add a new field `BackupTargetName` in `Spec` for a default backup target.

  ```golang
  type VolumeSpec struct {
    // The default backup target name when a backup is created from the volume.

>     BackupTargetName int64 `json:"backupTargetName"`

Reviewer 2: Should we handle this field during the upgrade for existing volumes? We might have overlooked this scenario in the implementation.

Author: We will handle it and we have the implementation to handle this scenario. Thanks

1. Prepare two clusters A and B with Longhorn installed.
2. Set up an extra backup target BT of two clusters with the same remote backup target.
3. Create the volume A and create a backup of the volume A in the cluster A.
4. In the cluster B, create a DR volume after the backup A is synchronized.
5. Write data to the volume A and create a new backup B in the cluster A.
6. Modifying/Deleting the default backup target URL is now allowed.
7. Check if the DR volume will synchronize the data in the cluster B.

#### Create And Restore A System Backup

1. Set up the default backup target and an extra backup target A.
2. Create a volume A with the default backup target name and a volume B with the extra backup target name A.
3. Write some data to volume A and volume B.
4. Create a system backup, and it succeeds. (the system backup and the volume A is stored on the default backup target, and the volume B is stored on the backup target A)
5. Delete the backup target A.
6. Restore the system backup, and it succeeds.
7. The backup target A should be restored.

#### Modify The Backup Target URL

1. Set up the default backup target and add an extra backup target A which has existing backups.

> 2. Existing backups on the extra backup target B can be synchronized back to the cluster.

Reviewer 5: ```suggestion
2. Existing backups on the extra backup target A can be synchronized back to the cluster.
```
I think they meant to say `Default backup` and `backup target-a` here.

Author: Modified. Thanks.
